### PR TITLE
feat: Prioritize accessible names check higher than inaccessibility check in `byRole` queries

### PR DESCRIPTION
**Pull request by kevin940726**
**What**:

Related to https://github.com/testing-library/dom-testing-library/issues/698 and https://github.com/testing-library/dom-testing-library/issues/820.

`*ByRole` queries are filtering inaccessibility (`hidden: false`) earlier than accessible names. This PR changes that by prioritizing accessible names over inaccessibility. Hopefully to achieve better performance in most cases.

**Why**:

As stated in https://github.com/testing-library/dom-testing-library/issues/820#issuecomment-726936225, one of the _escape hatches_ to improve the performance of the `*ByRole` queries is to pass `hidden: true` to the options. The reason is that the visibility check is more expensive in most cases. People are already using this technique to improve their queries' performance by setting the default options to true globally. The downside of this technique is obviously that we'd lose the visibility check for better and predictable queries.

What if instead of removing the inaccessibility check altogether, we just have to move the order lower? Since the query is really just a set of filter functions, if we can run the faster filter first and short-circuiting most of the candidates, then hopefully we don't have to run the slower inaccessibility check on so many elements.

For a naive test with 1000 buttons all with different names in jsdom environment, selecting the last element takes around 2 seconds. After applying this changes, it only takes around 1 second in average on my machine. It's not a lot but seems like a free win to me.

```js
const {getByRole} = render(
  Array.from({length: 1000})
    .map((_, index) => `<button>${index}</button>`)
    .join(''),
)
expect(getByRole('button', {name: '999'})).not.toBeNull()
```

Obviously it doesn't reflect the real world usage, but I'd expect the usage of accessible names check are more frequent than the inaccessibility check. It'd only be slower if there are multiple elements with the same name but only a part of them are accessible/visible. I can't think of a case where this would be a common scenario in testing. Additionally, people are already opting-out of the inaccessibility check globally, this should have little impact for them.

Preferably, it'd be nice to have it tested in some real world repos to see if it actually improves the performance though.

**How**:

Move the inaccessibility check lower to the last.

**Checklist**:

- [ ] Documentation added to the
      [docs site](https://github.com/testing-library/testing-library-docs) (I don't think this needs to be added to the documentation site. A record in changelog should be fine?)
- [ ] Tests (N/A)
- [ ] TypeScript definitions updated (N/A)
- [x] Ready to be merged

### CODE REVIEW

**Comment by kevin940726** [comment]
Any plans to merge this? Or should we get more approvals from maintainers?

**Comment by eps1lon** [comment]
Just wanted to let this sit so that everybody has a chance to review

**Review by timdeschryver — APPROVED** [review]
This LGTM. 
Although I agree with @markerikson that it could be made faster (even if it's just a little bit) if all `filter` statements are merged into 1 statement.

**Comment by markerikson** [comment]
Just as an FYI, I tried specifically adding a dep on `@testing-library/dom` to bump us to 8.11, and ran our longest UI test file, an "integration"-style test using `next-page-tester`.

Before:
```
 PASS  client/features/programDetails/characteristicsTab/tests/Characteristics.tests.tsx (107.118s)
  ProgramCharacteristicsContent
    √ Renders and allows editing Program Characteristics (98432ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        109.2s, estimated 124s
```

After:

```
 PASS  client/features/programDetails/characteristicsTab/tests/Characteristics.tests.tsx (90.958s)
  ProgramCharacteristicsContent
    √ Renders and allows editing Program Characteristics (82187ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        93.06s, estimated 106s
```

No other code changes.

So, this distinctly made _some_ improvement.

**Pull request merged**
